File: backend/app/schemas/vision.py
"""画像認識関連のAPIスキーマ

宿題プリント画像のアップロード・認識に使用するリクエスト/レスポンス定義。
"""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, Field


class RecognitionType(str, Enum):
    """認識タイプ"""

    HOMEWORK_PROBLEM = "homework_problem"
    HANDWRITING = "handwriting"
    DIAGRAM = "diagram"


class RecognizeImageRequest(BaseModel):
    """画像認識リクエスト"""

    image: str = Field(..., min_length=1, description="base64エンコードされた画像データ")
    recognition_type: RecognitionType = Field(
        default=RecognitionType.HOMEWORK_PROBLEM,
        description="認識タイプ",
    )
    expected_subject: str | None = Field(default=None, description="予想される教科")


class ProblemDetail(BaseModel):
    """認識された問題の詳細"""

    text: str = Field(..., description="問題文のテキスト")
    type: str = Field(
        ..., description="問題タイプ（arithmetic, word_problem, kanji, reading, other）"
    )
    difficulty: int = Field(..., description="難易度")
    expression: str | None = Field(default=None, description="計算式（算数の場合のみ）")


class RecognizeImageResponse(BaseModel):
    """画像認識成功レスポンス"""

    success: bool = Field(..., description="認識成功フラグ")
    problems: list[ProblemDetail] = Field(..., description="認識された問題リスト")
    confidence: float = Field(..., ge=0.0, le=1.0, description="認識の信頼度（0.0-1.0）")
    needs_confirmation: bool = Field(..., description="子供に確認が必要かどうか")


class RecognizeImageErrorResponse(BaseModel):
    """画像認識エラーレスポンス"""

    success: bool = Field(default=False, description="常にFalse")
    error_type: str = Field(
        ...,
        description="エラータイプ（recognition_failed, low_confidence, invalid_image, image_too_large）",
    )
    message: str = Field(..., description="エラーメッセージ")
    suggestions: list[str] = Field(default_factory=list, description="改善提案")

File: frontend/types/vision.ts
/**
 * カメラ・画像認識関連の型定義
 *
 * CameraInterfaceコンポーネントで使用するカメラ状態、エラー、
 * 画像認識リクエスト/レスポンスの型を定義。
 */

/** カメラの状態遷移 */
export type CameraStatus = "initial" | "active" | "preview" | "processing" | "recognized" | "error"

/**
 * カメラエラーの種別
 * - permission_denied: カメラ権限が拒否された
 * - not_available: カメラデバイスが利用不可
 * - capture_failed: 画像キャプチャに失敗
 * - recognition_failed: 画像認識に失敗
 * - unknown: その他のエラー
 */
export type CameraErrorType =
	| "permission_denied"
	| "not_available"
	| "capture_failed"
	| "recognition_failed"
	| "unknown"

/** カメラエラー情報 */
export interface CameraError {
	/** エラー種別 */
	type: CameraErrorType
	/** ユーザー向けエラーメッセージ（子供向けやさしい日本語） */
	message: string
}

/**
 * 画像認識リクエスト
 * バックエンド POST /api/v1/vision/recognize に送信
 */
export interface ImageRecognitionRequest {
	/** Base64エンコードされた画像データ */
	image: string
	/** 認識タイプ */
	recognition_type?: "homework_problem" | "handwriting" | "diagram"
	/** 予想される教科 */
	expected_subject?: string
}

/**
 * 認識された問題の詳細
 */
export interface ProblemDetail {
	/** 問題文のテキスト */
	text: string
	/** 問題タイプ（arithmetic, word_problem, kanji, reading, other） */
	type: string
	/** 難易度 */
	difficulty: number
	/** 計算式（算数の場合のみ） */
	expression?: string
}

/**
 * 画像認識レスポンス
 * バックエンドから返される認識結果
 */
export interface ImageRecognitionResponse {
	/** 認識成功フラグ */
	success: boolean
	/** 認識された問題リスト */
	problems: ProblemDetail[]
	/** 認識の確信度（0〜1） */
	confidence: number
	/** 子供に確認が必要かどうか */
	needs_confirmation: boolean
}
